File: 자료구조/Recursion/hanoi.java
package 자료구조.Recursion;

import java.util.ArrayList;

public class hanoi {
    ArrayList<int[]> result;
    public int[][] solution(int n) {
        result = new ArrayList<>(); //[[start,end]]
        han(1, 2, 3, n); // n개의 원판 1->2->3으로 이동
        int[][] answer = new int[result.size()][2]; //start, end 리스트 저장
        for (int i=0; i<result.size(); i++) {
            answer[i][0] = result.get(i)[0];
            answer[i][1] = result.get(i)[1];
        }
        return answer;
    }
    
    private void han(int start, int mid, int end, int n){
        int[] move = {start, end};
        if (n == 1) result.add(move); // 원판 하나면 그냥 이동
        else {
            han(start, end, mid, n-1); //start->mid 이동
            result.add(move); // 자신 1->3 바로 이동
            han(mid, start, end, n-1); //mid->end 이동
        }

    }
}

File: 자료구조/Recursion/hanoi.py
# n개의 원판을 1번째 기둥에서 3번째 기둥까지 옮기는 방법
# 기둥 3개, 원판 크기 작을수록 위에 쌓여져있음
# 한 번에 1 원판 이동 가능 & 큰 원판이 작은 원판 위에 오면 안됨
# n개의 원판을 최소 횟수로 이동할 것 (n은 15 이하)
# n=2 -> result: [[1,2],[1,3],[2,3]] #1->2,1->3,2->3 이동 있었다
# 기둥1의 남은 n-1개의 원판을 중간 기둥으로 이동(계속...마치 반복문)
# 기둥1의 n번째 원판 1개를 3번째 기둥으로 이동
# 기둥2의 남은 n-1개의 원판을 3번째 기둥으로 이동(계속...마치 반복문)
# 위의 과정을 n번 반복
def solution(n):
    answer = []
    def move(start, end, mid, idx): # start->end로 이동했음. mid는 나중에 옮길 곳
        if idx == 1: # 1개는 그냥 이동
            answer.append([start, end])
            return
        move(start, mid, end, idx-1) # 남은 원판 개수만큼 이동
        answer.append([start, end])
        move(mid, end, start, idx-1)
    move(1, 3, 2, n) # start, end, mid, idx
    return answer